File: components/offline_items_collection/core/android/offline_content_aggregator_bridge.cc
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "components/offline_items_collection/core/android/offline_content_aggregator_bridge.h"

#include <memory>
#include <utility>
#include <vector>

#include "base/android/callback_android.h"
#include "base/android/jni_string.h"
#include "base/bind.h"
#include "base/memory/ptr_util.h"
#include "components/offline_items_collection/core/android/offline_item_bridge.h"
#include "components/offline_items_collection/core/android/offline_item_share_info_bridge.h"
#include "components/offline_items_collection/core/android/offline_item_visuals_bridge.h"
#include "components/offline_items_collection/core/jni_headers/OfflineContentAggregatorBridge_jni.h"
#include "components/offline_items_collection/core/offline_item.h"
#include "components/offline_items_collection/core/throttled_offline_content_provider.h"

using base::android::AttachCurrentThread;
using base::android::ConvertJavaStringToUTF8;
using base::android::ConvertUTF8ToJavaString;
using base::android::JavaParamRef;
using base::android::ScopedJavaGlobalRef;
using base::android::ScopedJavaLocalRef;

namespace offline_items_collection {

using GetVisualsOptions = OfflineContentProvider::GetVisualsOptions;

namespace android {

namespace {
const char kOfflineContentAggregatorBridgeUserDataKey[] = "aggregator_bridge";

ContentId JNI_OfflineContentAggregatorBridge_CreateContentId(
    JNIEnv* env,
    const JavaParamRef<jstring>& j_namespace,
    const JavaParamRef<jstring>& j_id) {
  return ContentId(ConvertJavaStringToUTF8(env, j_namespace),
                   ConvertJavaStringToUTF8(env, j_id));
}

// Helper callback that glues the Java-specific callback logic to the native
// VisualsCallback that is required by the OfflineContentProvider native class.
void GetVisualsForItemHelperCallback(
    ScopedJavaGlobalRef<jobject> j_callback,
    const ContentId& id,
    std::unique_ptr<OfflineItemVisuals> visuals) {
  JNIEnv* env = AttachCurrentThread();
  Java_OfflineContentAggregatorBridge_onVisualsAvailable(
      env, j_callback, ConvertUTF8ToJavaString(env, id.name_space),
      ConvertUTF8ToJavaString(env, id.id),
      OfflineItemVisualsBridge::CreateOfflineItemVisuals(env,
                                                         std::move(visuals)));
}

void ForwardShareInfoToJavaCallback(
    ScopedJavaGlobalRef<jobject> j_callback,
    const ContentId& id,
    std::unique_ptr<OfflineItemShareInfo> shareInfo) {
  JNIEnv* env = AttachCurrentThread();
  Java_OfflineContentAggregatorBridge_onShareInfoAvailable(
      env, j_callback, ConvertUTF8ToJavaString(env, id.name_space),
      ConvertUTF8ToJavaString(env, id.id),
      OfflineItemShareInfoBridge::CreateOfflineItemShareInfo(
          env, std::move(shareInfo)));
}

void RenameItemCallback(ScopedJavaGlobalRef<jobject> j_callback,
                        RenameResult result) {
  base::android::RunIntCallbackAndroid(j_callback, static_cast<int>(result));
}

void RunGetAllItemsCallback(const base::android::JavaRef<jobject>& j_callback,
                            const std::vector<OfflineItem>& items) {
  JNIEnv* env = AttachCurrentThread();
  RunObjectCallbackAndroid(
      j_callback, OfflineItemBridge::CreateOfflineItemList(env, items));
}

void RunGetItemByIdCallback(const base::android::JavaRef<jobject>& j_callback,
                            const base::Optional<OfflineItem>& item) {
  JNIEnv* env = AttachCurrentThread();
  RunObjectCallbackAndroid(
      j_callback, item.has_value()
                      ? OfflineItemBridge::CreateOfflineItem(env, item.value())
                      : nullptr);
}

}  // namespace

// static
base::android::ScopedJavaLocalRef<jobject>
OfflineContentAggregatorBridge::GetBridgeForOfflineContentAggregator(
    OfflineContentAggregator* aggregator) {
  if (!aggregator->GetUserData(kOfflineContentAggregatorBridgeUserDataKey)) {
    aggregator->SetUserData(
        kOfflineContentAggregatorBridgeUserDataKey,
        base::WrapUnique(new OfflineContentAggregatorBridge(aggregator)));
  }
  OfflineContentAggregatorBridge* bridge =
      static_cast<OfflineContentAggregatorBridge*>(
          aggregator->GetUserData(kOfflineContentAggregatorBridgeUserDataKey));

  return ScopedJavaLocalRef<jobject>(bridge->java_ref_);
}

OfflineContentAggregatorBridge::OfflineContentAggregatorBridge(
    OfflineContentAggregator* aggregator)
    : provider_(std::make_unique<ThrottledOfflineContentProvider>(aggregator)) {
  JNIEnv* env = AttachCurrentThread();
  java_ref_.Reset(Java_OfflineContentAggregatorBridge_create(
      env, reinterpret_cast<intptr_t>(this)));

  provider_->AddObserver(this);
}

OfflineContentAggregatorBridge::~OfflineContentAggregatorBridge() {
  // TODO(dtrainor): Do not need to unregister because in the destructor of the
  // base class of OfflineContentAggregator.  Is |observers_| already dead?
  provider_->RemoveObserver(this);

  Java_OfflineContentAggregatorBridge_onNativeDestroyed(AttachCurrentThread(),
                                                        java_ref_);
}

void OfflineContentAggregatorBridge::OpenItem(
    JNIEnv* env,
    const JavaParamRef<jobject>& jobj,
    jint launch_location,
    const JavaParamRef<jstring>& j_namespace,
    const JavaParamRef<jstring>& j_id) {
  provider_->OpenItem(static_cast<LaunchLocation>(launch_location),
                      JNI_OfflineContentAggregatorBridge_CreateContentId(
                          env, j_namespace, j_id));
}

void OfflineContentAggregatorBridge::RemoveItem(
    JNIEnv* env,
    const JavaParamRef<jobject>& jobj,
    const JavaParamRef<jstring>& j_namespace,
    const JavaParamRef<jstring>& j_id) {
  provider_->RemoveItem(JNI_OfflineContentAggregatorBridge_CreateContentId(
      env, j_namespace, j_id));
}

void OfflineContentAggregatorBridge::CancelDownload(
    JNIEnv* env,
    const JavaParamRef<jobject>& jobj,
    const JavaParamRef<jstring>& j_namespace,
    const JavaParamRef<jstring>& j_id) {
  provider_->CancelDownload(JNI_OfflineContentAggregatorBridge_CreateContentId(
      env, j_namespace, j_id));
}

void OfflineContentAggregatorBridge::PauseDownload(
    JNIEnv* env,
    const JavaParamRef<jobject>& jobj,
    const JavaParamRef<jstring>& j_namespace,
    const JavaParamRef<jstring>& j_guid) {
  provider_->PauseDownload(JNI_OfflineContentAggregatorBridge_CreateContentId(
      env, j_namespace, j_guid));
}

void OfflineContentAggregatorBridge::ResumeDownload(
    JNIEnv* env,
    const JavaParamRef<jobject>& jobj,
    const JavaParamRef<jstring>& j_namespace,
    const JavaParamRef<jstring>& j_id,
    jboolean j_has_user_gesture) {
  provider_->ResumeDownload(JNI_OfflineContentAggregatorBridge_CreateContentId(
                                env, j_namespace, j_id),
                            j_has_user_gesture);
}

void OfflineContentAggregatorBridge::GetItemById(
    JNIEnv* env,
    const JavaParamRef<jobject>& jobj,
    const JavaParamRef<jstring>& j_namespace,
    const JavaParamRef<jstring>& j_id,
    const JavaParamRef<jobject>& jcallback) {
  OfflineContentProvider::SingleItemCallback callback =
      base::BindOnce(&RunGetItemByIdCallback,
                     base::android::ScopedJavaGlobalRef<jobject>(jcallback));
  provider_->GetItemById(JNI_OfflineContentAggregatorBridge_CreateContentId(
                             env, j_namespace, j_id),
                         std::move(callback));
}

void OfflineContentAggregatorBridge::GetAllItems(
    JNIEnv* env,
    const JavaParamRef<jobject>& jobj,
    const JavaParamRef<jobject>& jcallback) {
  OfflineContentProvider::MultipleItemCallback callback =
      base::BindOnce(&RunGetAllItemsCallback,
                     base::android::ScopedJavaGlobalRef<jobject>(jcallback));

  provider_->GetAllItems(std::move(callback));
}

void OfflineContentAggregatorBridge::GetVisualsForItem(
    JNIEnv* env,
    const JavaParamRef<jobject>& jobj,
    const JavaParamRef<jstring>& j_namespace,
    const JavaParamRef<jstring>& j_id,
    const JavaParamRef<jobject>& j_callback) {
  provider_->GetVisualsForItem(
      JNI_OfflineContentAggregatorBridge_CreateContentId(env, j_namespace,
                                                         j_id),
      GetVisualsOptions::IconOnly(),
      base::BindOnce(&GetVisualsForItemHelperCallback,
                     ScopedJavaGlobalRef<jobject>(env, j_callback)));
}

void OfflineContentAggregatorBridge::GetShareInfoForItem(
    JNIEnv* env,
    const JavaParamRef<jobject>& jobj,
    const JavaParamRef<jstring>& j_namespace,
    const JavaParamRef<jstring>& j_id,
    const JavaParamRef<jobject>& j_callback) {
  provider_->GetShareInfoForItem(
      JNI_OfflineContentAggregatorBridge_CreateContentId(env, j_namespace,
                                                         j_id),
      base::BindOnce(&ForwardShareInfoToJavaCallback,
                     ScopedJavaGlobalRef<jobject>(env, j_callback)));
}

void OfflineContentAggregatorBridge::RenameItem(
    JNIEnv* env,
    const JavaParamRef<jobject>& jobj,
    const JavaParamRef<jstring>& j_namespace,
    const JavaParamRef<jstring>& j_id,
    const JavaParamRef<jstring>& j_name,
    const JavaParamRef<jobject>& j_callback) {
  base::OnceCallback<void(RenameResult)> callback = base::BindOnce(
      &RenameItemCallback,
      base::android::ScopedJavaGlobalRef<jobject>(env, j_callback));

  provider_->RenameItem(JNI_OfflineContentAggregatorBridge_CreateContentId(
                            env, j_namespace, j_id),
                        ConvertJavaStringToUTF8(env, j_name),
                        std::move(callback));
}

void OfflineContentAggregatorBridge::OnItemsAdded(
    const OfflineContentProvider::OfflineItemList& items) {
  if (java_ref_.is_null())
    return;

  JNIEnv* env = AttachCurrentThread();
  Java_OfflineContentAggregatorBridge_onItemsAdded(
      env, java_ref_, OfflineItemBridge::CreateOfflineItemList(env, items));
}

void OfflineContentAggregatorBridge::OnItemRemoved(const ContentId& id) {
  if (java_ref_.is_null())
    return;

  JNIEnv* env = AttachCurrentThread();
  Java_OfflineContentAggregatorBridge_onItemRemoved(
      env, java_ref_, ConvertUTF8ToJavaString(env, id.name_space),
      ConvertUTF8ToJavaString(env, id.id));
}

void OfflineContentAggregatorBridge::OnItemUpdated(
    const OfflineItem& item,
    const base::Optional<UpdateDelta>& update_delta) {
  if (java_ref_.is_null())
    return;

  JNIEnv* env = AttachCurrentThread();
  Java_OfflineContentAggregatorBridge_onItemUpdated(
      env, java_ref_, OfflineItemBridge::CreateOfflineItem(env, item),
      OfflineItemBridge::CreateUpdateDelta(env, update_delta));
}

}  // namespace android
}  // namespace offline_items_collection

File: components/offline_items_collection/core/android/java/src/org/chromium/components/offline_items_collection/OfflineContentAggregatorBridge.java
// Copyright 2017 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.components.offline_items_collection;

import org.jni_zero.CalledByNative;
import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.base.Callback;
import org.chromium.base.ObserverList;
import org.chromium.build.annotations.NullMarked;
import org.chromium.build.annotations.Nullable;

import java.util.ArrayList;

/**
 * A helper class responsible for exposing the C++ OfflineContentAggregator
 * (components/offline_items_collection/core/offline_content_aggregator.h) class to Java.  This
 * class is created and owned by it's C++ counterpart OfflineContentAggregatorBridge
 * (components/offline_items_collection/core/android/offline_content_aggregator_bridge.h).
 */
@JNINamespace("offline_items_collection::android")
@NullMarked
public class OfflineContentAggregatorBridge implements OfflineContentProvider {
    private long mNativeOfflineContentAggregatorBridge;
    private final ObserverList<OfflineContentProvider.Observer> mObservers;

    /**
     * A private constructor meant to be called by the C++ OfflineContentAggregatorBridge.
     * @param nativeOfflineContentAggregatorBridge A C++ pointer to the
     * OfflineContentAggregatorBridge.
     */
    private OfflineContentAggregatorBridge(long nativeOfflineContentAggregatorBridge) {
        mNativeOfflineContentAggregatorBridge = nativeOfflineContentAggregatorBridge;
        mObservers = new ObserverList<OfflineContentProvider.Observer>();
    }

    // OfflineContentProvider implementation.
    @Override
    public void openItem(OpenParams openParams, ContentId id) {
        if (mNativeOfflineContentAggregatorBridge == 0) return;
        OfflineContentAggregatorBridgeJni.get()
                .openItem(
                        mNativeOfflineContentAggregatorBridge,
                        openParams.location,
                        openParams.openInIncognito,
                        id.namespace,
                        id.id);
    }

    @Override
    public void removeItem(ContentId id) {
        if (mNativeOfflineContentAggregatorBridge == 0) return;
        OfflineContentAggregatorBridgeJni.get()
                .removeItem(mNativeOfflineContentAggregatorBridge, id.namespace, id.id);
    }

    @Override
    public void cancelDownload(ContentId id) {
        if (mNativeOfflineContentAggregatorBridge == 0) return;
        OfflineContentAggregatorBridgeJni.get()
                .cancelDownload(mNativeOfflineContentAggregatorBridge, id.namespace, id.id);
    }

    @Override
    public void pauseDownload(ContentId id) {
        if (mNativeOfflineContentAggregatorBridge == 0) return;
        OfflineContentAggregatorBridgeJni.get()
                .pauseDownload(mNativeOfflineContentAggregatorBridge, id.namespace, id.id);
    }

    @Override
    public void resumeDownload(ContentId id) {
        if (mNativeOfflineContentAggregatorBridge == 0) return;
        OfflineContentAggregatorBridgeJni.get()
                .resumeDownload(mNativeOfflineContentAggregatorBridge, id.namespace, id.id);
    }

    @Override
    public void validateDangerousDownload(ContentId id) {
        if (mNativeOfflineContentAggregatorBridge == 0) return;
        OfflineContentAggregatorBridgeJni.get()
                .validateDangerousDownload(
                        mNativeOfflineContentAggregatorBridge, id.namespace, id.id);
    }

    @Override
    public void getItemById(ContentId id, Callback<OfflineItem> callback) {
        if (mNativeOfflineContentAggregatorBridge == 0) return;
        OfflineContentAggregatorBridgeJni.get()
                .getItemById(mNativeOfflineContentAggregatorBridge, id.namespace, id.id, callback);
    }

    @Override
    public void getAllItems(Callback<ArrayList<OfflineItem>> callback) {
        if (mNativeOfflineContentAggregatorBridge == 0) return;
        OfflineContentAggregatorBridgeJni.get()
                .getAllItems(mNativeOfflineContentAggregatorBridge, callback);
    }

    @Override
    public void getVisualsForItem(ContentId id, VisualsCallback callback) {
        OfflineContentAggregatorBridgeJni.get()
                .getVisualsForItem(
                        mNativeOfflineContentAggregatorBridge, id.namespace, id.id, callback);
    }

    @Override
    public void getShareInfoForItem(ContentId id, ShareCallback callback) {
        OfflineContentAggregatorBridgeJni.get()
                .getShareInfoForItem(
                        mNativeOfflineContentAggregatorBridge, id.namespace, id.id, callback);
    }

    @Override
    public void renameItem(ContentId id, String name, Callback</*RenameResult*/ Integer> callback) {
        OfflineContentAggregatorBridgeJni.get()
                .renameItem(
                        mNativeOfflineContentAggregatorBridge, id.namespace, id.id, name, callback);
    }

    @Override
    public void addObserver(final OfflineContentProvider.Observer observer) {
        mObservers.addObserver(observer);
    }

    @Override
    public void removeObserver(OfflineContentProvider.Observer observer) {
        mObservers.removeObserver(observer);
    }

    // Methods called from C++ via JNI.
    @CalledByNative
    private void onItemsAdded(ArrayList<OfflineItem> items) {
        if (items.size() == 0) return;

        for (Observer observer : mObservers) {
            observer.onItemsAdded(items);
        }
    }

    @CalledByNative
    private void onItemRemoved(String nameSpace, String id) {
        ContentId contentId = new ContentId(nameSpace, id);

        for (Observer observer : mObservers) {
            observer.onItemRemoved(contentId);
        }
    }

    @CalledByNative
    private void onItemUpdated(OfflineItem item, UpdateDelta updateDelta) {
        for (Observer observer : mObservers) {
            observer.onItemUpdated(item, updateDelta);
        }
    }

    @CalledByNative
    private static void onVisualsAvailable(
            VisualsCallback callback, String nameSpace, String id, OfflineItemVisuals visuals) {
        callback.onVisualsAvailable(new ContentId(nameSpace, id), visuals);
    }

    @CalledByNative
    private static void onShareInfoAvailable(
            ShareCallback callback, String nameSpace, String id, OfflineItemShareInfo shareInfo) {
        callback.onShareInfoAvailable(new ContentId(nameSpace, id), shareInfo);
    }

    /**
     * Called when the C++ OfflineContentAggregatorBridge is destroyed.  This tears down the Java
     * component of the JNI bridge so that this class, which may live due to other references, no
     * longer attempts to access the C++ side of the bridge.
     */
    @CalledByNative
    private void onNativeDestroyed() {
        mNativeOfflineContentAggregatorBridge = 0;
    }

    /**
     * A private static factory method meant to be called by the C++ OfflineContentAggregatorBridge.
     * @param nativeOfflineContentAggregatorBridge A C++ pointer to the
     * OfflineContentAggregatorBridge.
     * @return A new instance of this OfflineContentAggregatorBridge.
     */
    @CalledByNative
    private static OfflineContentAggregatorBridge create(
            long nativeOfflineContentAggregatorBridge) {
        return new OfflineContentAggregatorBridge(nativeOfflineContentAggregatorBridge);
    }

    @NativeMethods
    interface Natives {
        void openItem(
                long nativeOfflineContentAggregatorBridge,
                @LaunchLocation int location,
                boolean openInIncognito,
                @Nullable String nameSpace,
                @Nullable String id);

        void removeItem(
                long nativeOfflineContentAggregatorBridge,
                @Nullable String nameSpace,
                @Nullable String id);

        void cancelDownload(
                long nativeOfflineContentAggregatorBridge,
                @Nullable String nameSpace,
                @Nullable String id);

        void pauseDownload(
                long nativeOfflineContentAggregatorBridge,
                @Nullable String nameSpace,
                @Nullable String id);

        void resumeDownload(
                long nativeOfflineContentAggregatorBridge,
                @Nullable String nameSpace,
                @Nullable String id);

        void validateDangerousDownload(
                long nativeOfflineContentAggregatorBridge,
                @Nullable String nameSpace,
                @Nullable String id);

        void getItemById(
                long nativeOfflineContentAggregatorBridge,
                @Nullable String nameSpace,
                @Nullable String id,
                Callback<OfflineItem> callback);

        void getAllItems(
                long nativeOfflineContentAggregatorBridge,
                Callback<ArrayList<OfflineItem>> callback);

        void getVisualsForItem(
                long nativeOfflineContentAggregatorBridge,
                @Nullable String nameSpace,
                @Nullable String id,
                VisualsCallback callback);

        void getShareInfoForItem(
                long nativeOfflineContentAggregatorBridge,
                @Nullable String nameSpace,
                @Nullable String id,
                ShareCallback callback);

        void renameItem(
                long nativeOfflineContentAggregatorBridge,
                @Nullable String nameSpace,
                @Nullable String id,
                String name,
                Callback</*RenameResult*/ Integer> callback);
    }
}
